[STAGE-3] Security

polinatonko opened the pull request:
Complete below steps:
1. Store user details in memory for Basic Auth flow
2. Secure API endpoints locally and allow access for all authenticated users (Basic Auth)
3. Configure security for Spring Actuator endpoints locally and allow access for MANAGER role only (Basic Auth)
4. Secure API endpoints in cloud using XSUAA service and allow access for all authenticated users (OAuth2)
5. Configure security for Spring Actuator endpoints in cloud (Basic Auth) and allow access for MANAGER role only
6. Run PostgreSQL in Docker and switch to it locally (H2 is removed)
7. Replace manifest.yaml with mta.yaml and redeploy application
8. Add authentication tests, switch to PostgreSQL containers in the tests

polinatonko committed:
chore: set up initial project structure

polinatonko committed:
chore: reformat pom.xml, remove odd lines from .gitignore, switch to yaml format of properties files

polinatonko committed:
Merge pull request #1 from polinatonko/feature/init-project

[STAGE-1] Init project backbone

polinatonko committed:
chore: set up database schema using Liquibase, configure H2 database for local profile

polinatonko committed:
Merge pull request #2 from polinatonko/feature/set-up-db-schema

[STAGE-1] Set up database schema and configure H2 database

polinatonko committed:
chore: update course schema, fix one-to-one course_settings mapping

polinatonko committed:
feat: implement basic crud functionality for students and courses

polinatonko committed:
feat: add global exception handler and dto to represent api errors

polinatonko committed:
docs: add swagger for rest api

polinatonko committed:
feat: add validation for student and course dtos, improve exception handling related to client errors

polinatonko committed:
refactor: rename fields, class methods and dtos; add 'toDtos' method to mappers

polinatonko committed:
refactor: fix typos and rename the application in the properties and docs

polinatonko committed:
refactor: add 'toDto' method in exception handler, use static imports for http statuses

polinatonko committed:
refactor: reverse one-to-one relationship, add enrollment entity to represent many-to-many relationship; rename liquibase scripts

polinatonko committed:
refactor: make minor changes

polinatonko committed:
feat: enable spring boot actuator (health, info endpoints)

polinatonko committed:
feat: disable open-in-view feature; add logic to daily collect information about upcoming courses and send email notifications to enrolled students

polinatonko committed:
refactor: refactor dtos

polinatonko committed:
feat: add basic crud functionality for lessons

polinatonko committed:
feat: implement logic for coin-based payment

polinatonko committed:
test: add integration and unit tests for enrollment functionality

polinatonko committed:
chore: add test profile

polinatonko committed:
refactor: apply best practices

polinatonko committed:
feat: add loggings of exceptions, add validator for lesson; fix: creating and updating lessons

polinatonko committed:
refactor: change project structure to package by feature

polinatonko committed:
refactor: modify project structure, add loading application version from pom.xml

polinatonko committed:
feat: configure manifest.yaml

polinatonko committed:
feat: bind hana db, application logging service and feature flags service to the application

polinatonko committed:
chore: remove default values for uuid

polinatonko committed:
chore: remove default values for uuid

polinatonko committed:
refactor: replace literals with constants, add missing annotations and remove unnecessary ones, rename variables

polinatonko committed:
chore: change default liquibase uuid type for sap hana db to varbinary

polinatonko committed:
refactor: replace the use of mailtrap sdk with spring mails; add spring retry dependency

polinatonko committed:
feat: add retrieving destinations from the destination service

polinatonko committed:
refactor: replace string literals with constants, add retryable and retry annotations for http calls to the api of the feature flags service

polinatonko committed:
feat: add bindings to the sap btp services, enable remote debugging

polinatonko committed:
refactor: refactor job for sending course notifications

polinatonko committed:
Merge branch 'stage-1' into stage-2

polinatonko committed:
feat: add caching to the entity management services

polinatonko committed:
chore: update version of the project

polinatonko committed:
refactor: refactor destination, feature flags and email services

polinatonko committed:
refactor: transfer the logic of course enrollment validation to the validator class, update the unit-test accordingly

polinatonko committed:
Merge branch 'stage-1' into stage-2

polinatonko committed:
refactor: remove manual creation of authentication headers

polinatonko committed:
refactor: add btp package for btp related logic, use static imports for well-known constants, remove unnecessary dependency version declaration from pom.xml

polinatonko committed:
style: remove blank lines from the changelog file

polinatonko committed:
refactor: refactor stage-2

polinatonko committed:
refactor: rename methods, use static imports

polinatonko committed:
refactor: fix method's naming, imports, edit pom.xml

polinatonko committed:
chore: edit .gitignore

polinatonko committed:
feat: set up spring security, add basic authentication locally

polinatonko committed:
feat: add authentication for /actuator endpoints (manager role), authentication exception handling

polinatonko committed:
feat: secure api endpoints in cloud using xsuaa

polinatonko committed:
feat: run postgresql in docker, switch to it locally

polinatonko committed:
refactor: rewrite the destination mapping, destination and token services, rest clientconfiguration, method names

polinatonko committed:
feat: add mta.yaml file, edit docker-compose and security configurations, switch to postgres locally

polinatonko committed:
fix: lost update when enrolling in a course

polinatonko committed:
Merge branch 'stage-2' into stage-3

polinatonko committed:
chore: use postgresql containers in tests

polinatonko committed:
style: add empty lines between class fields, reorganize lines of the code

polinatonko committed:
test: add authentication tests

polinatonko committed:
chore: restore manifest.yaml

SN4NTR commented on src/test/java/org/example/learningsystem/course/SecurityTest.java:
```diff
@@ -0,0 +1,97 @@
+    private MockMvc mockMvc;
+
+    @DynamicPropertySource
+    static void configure(DynamicPropertyRegistry registry) {
+        registry.add("spring.datasource.url", POSTGRESQL_CONTAINER::getJdbcUrl);
+        registry.add("spring.datasource.username", POSTGRESQL_CONTAINER::getUsername);
+        registry.add("spring.datasource.password", POSTGRESQL_CONTAINER::getPassword);
+    }
```
it's better to do not depend on credentials generated by testcontainers, but define your own in `application-test.yaml` file and then use them to build container:
```
@Container
    private static final PostgreSQLContainer<?> POSTGRESQL_CONTAINER = new PostgreSQLContainer<>("postgres:17.4")
            .withUsername()
            .withPassword();
```

SN4NTR commented on pom.xml:
```diff
@@ -23,6 +23,7 @@
 		<git.commit-id.version>9.0.1</git.commit-id.version>
 		<ngdbc.version>2.24.7</ngdbc.version>
 		<liquibase-hanadb.version>4.31.1</liquibase-hanadb.version>
+		<sap.cloud.security.version>3.6.0</sap.cloud.security.version>
```
please also increase your application version to 1.2.0

polinatonko committed:
refactor: refactor stage-3

Use volume for PostgreSQL in the docker-compose.yaml. Add the health
check parameters to the mta.yaml. Allow unauthenticated access to the
/actuator/health endpoint. Update the project version. Add static imports.
Add the basic authentication credentials to the application.yaml. Update
the definitions of filter chains inside the security configuration classes.
Update the usage of testcontainers and enable security filters in the tests.

polinatonko committed:
refactor: fix names of classes and methods, use string constants, remove method chains in handlers, add empty lines

polinatonko committed:
refactor: use record type for user credentials

polinatonko committed:
Merge branch 'main' into stage-3
